### PR TITLE
Fix Decimal integration test failures

### PR DESCRIPTION
**Pull request by razajafri**
This PR addresses most of the test failures in #6988 

The only ones remaining are due to the overflow, which will be coming up shortly. 

This PR includes 
* Fixing of exception messages to match Spark 3.4
* Skipping a test param with negative scale because of a bug introduced as a result of https://github.com/apache/spark/commit/301a13963808d1ad44be5cacf0a20f65b853d5a2 in Spark. I have filed an issue in Spark https://issues.apache.org/jira/browse/SPARK-41207
* Some expressions are now required to have a `resultDecimalType` since expressions are now responsible for handling the inputs and outputs for Decimal types

### CODE REVIEW

File: sql-plugin/src/main/340+/scala/org/apache/spark/sql/rapids/arithmetic.scala [hunk 1]
```diff
@@ -0,0 +1,491 @@
...
+  override def inputType: AbstractDataType = TypeCollection.NumericAndInterval
+
+  override def symbol: String = "+"
+
+  override def binaryOp: BinaryOp = BinaryOp.ADD
+
+  override def astOperator: Option[BinaryOperator] = Some(ast.BinaryOperator.ADD)
+
+  override def doColumnar(lhs: BinaryOperable, rhs: BinaryOperable): ColumnVector = {
+    val ret = super.doColumnar(lhs, rhs)
```
Comment (revans2) [inline sql-plugin/src/main/340+/scala/org/apache/spark/sql/rapids/arithmetic.scala:76]: I thought that we needed to do the add as something larger than a 128-bit value. Is that not true? Do we just not have tests that cover these odd use cases?  Like add Decimal(30, 0) to a Decimal(30, 10)? This also goes for subtract too...

File: sql-plugin/src/main/311until340-all/scala/org/apache/spark/sql/rapids/arithmetic.scala [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+      if (dataType.isInstanceOf[DecimalType]) {
+        decimalOpOverflowCheck(lhs, rhs, ret)
+      } else {
+        ret.incRefCount()
+      }
+    }
+  }
+}
+
+case class GpuMultiply(
```
Comment (revans2) [inline sql-plugin/src/main/311until340-all/scala/org/apache/spark/sql/rapids/arithmetic.scala:172]: Why do we shim GpuMultiply if it does not handle decimal at all . Did something else change that makes it so we need to have multiple copies of it?

File: sql-plugin/src/main/311until340-all/scala/org/apache/spark/sql/rapids/arithmetic.scala [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+
+  override def inputType: AbstractDataType = NumericType
+
+  override def symbol: String = "*"
+
+  override def binaryOp: BinaryOp = BinaryOp.MUL
+  override def astOperator: Option[BinaryOperator] = Some(ast.BinaryOperator.MUL)
+}
+
+case class GpuDivide(left: Expression, right: Expression,
```
Comment (revans2) [inline sql-plugin/src/main/311until340-all/scala/org/apache/spark/sql/rapids/arithmetic.scala:186]: Same for GpuDivide. Did something change and we need multiple non-decimal versions of it?

File: sql-plugin/src/main/311until340-all/scala/org/apache/spark/sql/rapids/arithmetic.scala [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+  override def inputType: AbstractDataType = TypeCollection(DoubleType, DecimalType)
+
+  override def symbol: String = "/"
+
+  override def binaryOp: BinaryOp = BinaryOp.TRUE_DIV
+
+  override def outputTypeOverride: DType = GpuColumnVector.getNonNestedRapidsType(dataType)
+}
+
+case class GpuIntegralDivide(left: Expression, right: Expression) extends GpuDivModLike {
```
Comment (revans2) [inline sql-plugin/src/main/311until340-all/scala/org/apache/spark/sql/rapids/arithmetic.scala:203]: Do any of these actually need to be in the Shim?

File: sql-plugin/src/main/340+/scala/org/apache/spark/sql/rapids/arithmetic.scala [hunk 5]
```diff
@@ -0,0 +1,491 @@
...
+  override def astOperator: Option[BinaryOperator] = Some(ast.BinaryOperator.MUL)
+
+  // scalastyle:off
+  // The formula follows Hive which is based on the SQL standard and MS SQL:
+  // https://cwiki.apache.org/confluence/download/attachments/27362075/Hive_Decimal_Precision_Scale_Support.pdf
+  // https://msdn.microsoft.com/en-us/library/ms190476.aspx
+  // Result Precision: p1 + p2 + 1
+  // Result Scale:     s1 + s2
+  // scalastyle:on
+  override def resultDecimalType(p1: Int, s1: Int, p2: Int, s2: Int): DecimalType = {
```
Comment (revans2) [inline sql-plugin/src/main/340+/scala/org/apache/spark/sql/rapids/arithmetic.scala:256]: If GpuDecimalMultiply handles the decimal values, then why do we need these changes here at all?

**Review by revans2 — COMMENTED** [review]
I started to go through this, but I didn't finish. I am just really confused about a lot of the decisions on what to shim. We have GpuMultiply and GpuDivide that do not support decimal but are shimmed as a part of this. We have GpuAdd and GpuSubtract that handle decimal values, but the shims do not look like they will work in all cases. I really would like to see these addressed and also have a lot of tests added where the LHS and RHS have very different scales where the resulting precision would be 38 and require the precision and scale to be adjusted. We did not test these use cases originally because we knew that Spark would make the LHS and RHS the same type before processing, which meant we didn't need to test all of the various types.

File: sql-plugin/src/main/311until340-all/scala/org/apache/spark/sql/rapids/arithmetic.scala [hunk 6]
```diff
@@ -0,0 +1,246 @@
...
+    right: Expression,
+    failOnError: Boolean) extends CudfBinaryArithmetic {
+  override def inputType: AbstractDataType = TypeCollection.NumericAndInterval
+
+  override def symbol: String = "-"
+
+  override def binaryOp: BinaryOp = BinaryOp.SUB
+  override def astOperator: Option[BinaryOperator] = Some(ast.BinaryOperator.SUB)
+
+  private[this] def basicOpOverflowCheck(
```
[the remark this answers is not in the document]
Comment (razajafri) [reply sql-plugin/src/main/311until340-all/scala/org/apache/spark/sql/rapids/arithmetic.scala:82]: The issue is still not fixed. I am going to either put up a follow-on PR or push updates to this if this isn't in soon

File: sql-plugin/src/main/340+/scala/org/apache/spark/sql/rapids/arithmetic.scala (hunk 5 above)
Comment (razajafri) [reply sql-plugin/src/main/340+/scala/org/apache/spark/sql/rapids/arithmetic.scala:256]: Thanks. This change isn't needed as we are handling the Decimal multiply in a separate class.

File: sql-plugin/src/main/340+/scala/org/apache/spark/sql/rapids/arithmetic.scala (hunk 1 above)
Comment (razajafri) [reply sql-plugin/src/main/340+/scala/org/apache/spark/sql/rapids/arithmetic.scala:76]: I will fix that as a separate PR

File: sql-plugin/src/main/311until340-all/scala/org/apache/spark/sql/rapids/arithmetic.scala (hunk 2 above)
Comment (razajafri) [reply sql-plugin/src/main/311until340-all/scala/org/apache/spark/sql/rapids/arithmetic.scala:172]: Sorry, this was an oversight and misunderstanding of how we are handling decimal multiply and divide

File: sql-plugin/src/main/311until340-all/scala/org/apache/spark/sql/rapids/arithmetic.scala (hunk 3 above)
Comment (razajafri) [reply sql-plugin/src/main/311until340-all/scala/org/apache/spark/sql/rapids/arithmetic.scala:186]: same comment as above

File: sql-plugin/src/main/311until340-all/scala/org/apache/spark/sql/rapids/arithmetic.scala (hunk 4 above)
Comment (razajafri) [reply sql-plugin/src/main/311until340-all/scala/org/apache/spark/sql/rapids/arithmetic.scala:203]: Yes, we need the `resultDecimalType` which will be called by the unsafeprojection if the expression outputs a DecimalType

File: sql-plugin/src/main/311until340-all/scala/org/apache/spark/sql/rapids/RapidsSchemaUtils.scala [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.rapids
+
+import org.apache.spark.sql.catalyst.analysis.Resolver
+import org.apache.spark.sql.types.StructType
+import org.apache.spark.sql.util.SchemaUtils
+
+object RapidsSchemaUtils {
```
Comment (revans2) [inline sql-plugin/src/main/311until340-all/scala/org/apache/spark/sql/rapids/RapidsSchemaUtils.scala:23]: If you are fixing a compile error like this, please pull it out into a separate PR.  This is something that will benefit everyone by unblocking the build and should not be combined with and blocked by an unrelated change going in.

**Comment by razajafri** [comment]
@revans2 will it be possible to merge this soon so we can get started on the databricks 11.3 shims?

What are some of the things you want me to do to get this in ASAP? This PR is following the `incremental improvement` towards the 3.4 as we have done in the past in this effort

File: sql-plugin/src/main/311until340-all/scala/org/apache/spark/sql/rapids/RapidsSchemaUtils.scala (hunk 7 above)
Comment (revans2) [reply sql-plugin/src/main/311until340-all/scala/org/apache/spark/sql/rapids/RapidsSchemaUtils.scala:23]: Is this still needed now that https://github.com/NVIDIA/spark-rapids/pull/7153 is merged in?

File: sql-plugin/src/main/340+/scala/org/apache/spark/sql/rapids/arithmetic.scala (hunk 1 above)
Comment (revans2) [reply sql-plugin/src/main/340+/scala/org/apache/spark/sql/rapids/arithmetic.scala:72]: Then please add a comment explaining that even if this does fix the current test failures this still has a bug in it and point to that issue.

File: sql-plugin/src/main/340+/scala/org/apache/spark/sql/rapids/arithmetic.scala [hunk 8]
```diff
@@ -0,0 +1,384 @@
...
+}
+
+trait GpuDivModLikeWithPromote extends GpuDivModLike with DecimalWithPromote
+
+trait DecimalWithPromote extends CudfBinaryOperator {
+  private def getCastedVectorIfNeeded(
+      vector: GpuColumnVector,
+      d: DecimalType, c: DType): GpuColumnVector = {
+    if (!vector.dataType.sameType(d)) {
+      GpuColumnVector.from(vector.getBase.castTo(c), d)
```
[the remark this answers is not in the document]
Comment (revans2) [reply sql-plugin/src/main/340+/scala/org/apache/spark/sql/rapids/arithmetic.scala:181]: I still don't understand why we are doing this for integral divide?

**Review by revans2 — COMMENTED** [review]
It looks good, I am mostly confused about integral divide now.

File: sql-plugin/src/main/340+/scala/org/apache/spark/sql/rapids/arithmetic.scala (hunk 8 above)
[the remark this answers is not in the document]
Comment (razajafri) [reply sql-plugin/src/main/340+/scala/org/apache/spark/sql/rapids/arithmetic.scala:181]: As I understand the code in IntegralDivide didn't have to promote precision as it was being done in `DecimalPrecision.scala` [here](https://github.com/apache/spark/commit/301a13963808d1ad44be5cacf0a20f65b853d5a2?diff=unified#diff-a3c51d44e59e904c5ac9f75b2abc49959181145649dcc51c2437f7fbc7bb5be7L186). Now if we don't do this, the resultDecimalType will return a DecimalType that may not match the type of the inputs. Hence we have to cast them to the expected type. Otherwise, the implicit conversion fails, now that I think about this again, we could try to override the outputType. Let me try that

[the remark this answers is not in the document]
Comment (razajafri) [reply sql-plugin/src/main/340+/scala/org/apache/spark/sql/rapids/arithmetic.scala:181]: Yeah, that doesn't work either as that is just deciding what the outputType should be, the inputs still have to be casted.

[the remark this answers is not in the document]
Comment (revans2) [reply sql-plugin/src/main/340+/scala/org/apache/spark/sql/rapids/arithmetic.scala:181]: My main concern is that we get the correct answer.  From what I have seen the new Integral divide code for decimals does the divide using the original values, but the output scale is always set to 0, no matter what the precision is. Then the result is cast to a long. I don't think that corresponds to what this patch is doing. 

For example what happens if we do `1 DIV 0.5` The answer should be `2`. But what happens if the decimal types make it so we lose information if we do the cast ahead of time?  `CAST(1 AS DECIMAL(38, 0)) DIV CAST(0.5 AS DECIMAL(38, 38))`. The answer on older versions of Spark is null because the cast of 1 to decimal(38,38) overflowed.  Your new code does not have the overflow checking. I am not sure what the cast is going to do here, and I don't know what the answer is going to be. I think in 3.4 the answer will still be 2 because it does the divide prior to any other operations, and then figures out how to make it a long.

[the remark this answers is not in the document]
Comment (razajafri) [reply sql-plugin/src/main/340+/scala/org/apache/spark/sql/rapids/arithmetic.scala]: I have repurposed the `DecimalDivide` and used it for `IntegralDivide` I've had to modify the divide128 kernel for it as the regular divide rounds up the quotient.You can look at the cudf PR [here](https://github.com/NVIDIA/spark-rapids-jni/pull/803)

**Comment by razajafri** [comment]
> It looks good, I am mostly confused about integral divide now.

I have updated the patch to now use the `GpuDecimalDivide` which does decimal division in 128-bit, I then return the quotient without rounding it. This gives us a long based on the original quotient. One thing that is very odd in Spark is that the overflow is still based on the original 128-bit Decimal quotient and not the Long value.
Can you look at this again please

**Comment by razajafri** [comment]
I am converting this to a draft as there are only a couple of test failures left and there's no point in letting this go in at this point. This has become a huge PR and so in an effort to make it easier for the reviewers I will be breaking this into smaller PRs

**Comment by razajafri** [comment]
This PR is closed in favor of https://github.com/NVIDIA/spark-rapids/pull/7687
